File: geekbang/algorithm-models/8.1- KMP 字符串匹配模板/strStr.go
package main

func strStr(haystack string, needle string) int {

	n, m := len(haystack), len(needle)
	if m == 0 {
		return 0
	}
	next := make([]int, m)
	for i, j := 1, 0; i < m; i++ {
		for j > 0 && needle[i] != needle[j] {
			j = next[j-1]
		}
		if needle[i] == needle[j] {
			j++
		}

		next[i] = j
	}
	for i, j := 0, 0; i < n; i++ {
		for j > 0 && needle[j] != haystack[i] {
			j = next[j-1]
		}
		if needle[j] == haystack[i] {
			j++
		}
		if j == m {
			return i - m + 1
		}

	}

	return -1
}

File: geekbang/algorithm-models/8.1- KMP 字符串匹配模板/strStr.js
// JavaScript
// LeetCode 28 实现strStr
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function(haystack, needle) {
        if (needle.length == 0) return 0;
        const n = haystack.length;
        const m = needle.length;
        let next = [];
        for (let i = 0; i < m; i++) next[i] = -1; // 下标从0开始，初值-1；下标从1开始，初值0。
        for (let i = 1, j = -1; i < m; i++) {
            while (j >= 0 && needle[j + 1] != needle[i]) j = next[j];
            if (needle[j + 1] == needle[i]) j++;
            next[i] = j;
        }
        for (let i = 0, j = -1; i < n; i++) {
            while (j >= 0 && needle[j + 1] != haystack[i]) j = next[j];
            if (j + 1 < m && needle[j + 1] == haystack[i]) j++;
            if (j + 1 == m) return i - m + 1;
        }
        return -1;
};
